src/context.jl

```diff
@@ -1,105 +1,230 @@
-* `ctx::Context`: the context to pass into `ismissing`
+* `ctx::AbstractContext`: the context to pass into `ismissing`
 * `data::AbstractVector`: the data array to search
 
 # Returns
 * `Int`: the next index in `data` that isn't missing
 """
-function Base.findnext(ctx::Context, data::AbstractVector, idx::Int)
+function Base.findnext(ctx::AbstractContext, data::AbstractVector, idx::Int)
     return findnext(x -> !ismissing(ctx, x), data, idx)
 end
+
+mutable struct Context <: AbstractContext
+    num::Int
+    count::Int
+    limit::Float64
+    is_missing::Function
+    on_complete::Function
+end
+
+"""
+    Context
+
+Records base information about the missing data and assume all observations are equally
+weighted.
+
+# Keyword Arguments
+* `n::Int`: number of observations
+* `count::Int`: number of missing values found
+* `limit::Float64`: portion of total values allowed to be imputed (should be between 0.0 and 1.0).
+* `is_missing::Function`: returns a Bool if the value counts as missing
+* `on_complete::Function`: a function to run when imputation is complete
+"""
+function Context(;
+    limit::Float64=0.1,
+    is_missing::Function=ismissing,
+    on_complete::Function=complete
+)
+    return Context(0, 0, limit, is_missing, on_complete)
+end
+
+function Base.empty(ctx::Context)
+    _ctx = copy(ctx)
+    _ctx.num = 0
+    _ctx.count = 0
+
+    return _ctx
+end
+
+function missing_update!(ctx::Context, was_missing)
+    ctx.num += 1
+
+    if was_missing
+        ctx.count += 1
+    end
+end
+
+function complete(ctx::Context, data)
+    missing_ratio = ctx.count / ctx.num
+
+    if missing_ratio > ctx.limit
+        throw(ImputeError(
+            "More than $(ctx.limit * 100)% of values were missing ($missing_ratio)."
+        ))
+    end
+
+    return data
+end
+
+
+mutable struct WeightedContext <: AbstractContext
+    num::Int
+    s::Float64
+    limit::Float64
+    is_missing::Function
+    on_complete::Function
+    wv::AbstractWeights
+end
+
+"""
+    WeightedContext(wv; limit=1.0, is_missing=ismissing, on_complete=complete)
+
+Records information about the missing data relative to a set of weights.
+This context type can be useful if some missing observation are more important than others
+(e.g., more recent observations in time series datasets)
+
+# Arguments
+* `wv::AbstractWeights`: a set of statistical weights to use when evaluating the importance
+  of each observation. Will be accumulated during imputation.
+
+# Keyword Arguments
+* `num::Int`: number of observations
+* `s::Float64`: sum of the weights of missing values
+* `limit::Float64`: portion of total values allowed to be imputed (should be between 0.0 and 1.0).
+* `is_missing::Function`: returns a Bool if the value counts as missing
+* `on_complete::Function`: a function to run when imputation is complete
+"""
+function WeightedContext(
+    wv::AbstractWeights;
+    limit::Float64=1.0,
+    is_missing::Function=ismissing,
+    on_complete::Function=complete
```

Comment:
I thought we decided we were getting rid of this?

Reply:
I think it'll be useful for handling custom imputation failure logic without needing a new context type. For example, you could change the `on_complete` function to throw a warning instead.

Reply:
I'm not extatic about it about it, but Ok